packages/plugin-node-surrounding-code/surrounding-code.js

```diff
@@ -28,16 +31,56 @@ module.exports = {
 }
 
 const getSurroundingCode = (file, lineNumber, cb) => {
-  readFile(file, 'utf8', (err, data) => {
-    if (err) return cb(err)
-    const lines = data.split('\n')
-    const start = Math.max(0, lineNumber - SURROUNDING_LINES - 1)
-    const end = Math.min(lines.length, lineNumber + SURROUNDING_LINES)
-    cb(null, lines
-      .slice(start, end)
-      .reduce((accum, line, i) => {
-        accum[start + (i + 1)] = line
-        return accum
-      }, {}))
+  const start = lineNumber - SURROUNDING_LINES
+  const end = lineNumber + SURROUNDING_LINES
+
+  const reader = createReadStream(file, { encoding: 'utf8' })
+  const splitter = new byline.LineStream({ keepEmptyLines: true })
+  const slicer = new CodeRange({ start, end })
+
+  // if the slicer has enough lines already, no need to keep reading from the file
+  slicer.on('done', () => reader.destroy())
+
+  pump(reader, splitter, slicer, (err) => {
+    // reader.destroy() causes a "premature close" error which we can tolerate
+    if (err && err.message !== 'premature close') return cb(err)
+    cb(null, slicer.getCode())
   })
 }
+
+// This writable stream takes { start, end } options specifying the
+// range of lines that should be extracted from a file. Pipe a readable
+// stream to it that provides source lines as each chunk. If the range
+// is satisfied before the end of the readable stream, it will emit the
+// 'done' event. Once a 'done' or 'finish' event has been seen, call getCode()
+// to get the range in the following format:
+// {
+//   '10': 'function getSquare (cb) {',
+//   '11': '  rectangles.find({',
+//   '12': '    length: 12',
+//   '13': '    width: 12',
+//   '14': '  }, err => cb)',
+//   '15': '}'
+// }
+class CodeRange extends Writable {
+  constructor (opts) {
+    super({ ...opts, decodeStrings: false })
```

Comment:
by default, writable streams convert strings into raw buffers but we don't want that here